[venuslimm] iP

venuslimm opened the pull request:
# Fake Duke
> I might be fake but my features are real!

## Welcome to Fake Duke 🧔 
Fake Duke is a Duke which can help you *remember all your tasks*. It's
- Easy to use
- Free of charge
- ***AN UPGRADED DUKE***

## Features 💯 
- [x] CLI version
- [x] Managing tasks
- [ ] GUI version

## How to use the latest Fake Duke? 🤔 
1. Download and setup [Java JRE](https://www.oracle.com/sg/java/technologies/downloads/) on your machine
2. Download the [JAR file](https://github.com/venuslimm/ip/releases/download/A-Jar/duke.jar) and double click on your machine. 
2.1. Or you can choose to run this command in your terminal: `java -jar <jar_file_name>.jar`
3. Do your tasks management!

### Actions on Fake Duke
- Add task
  - Todo: `todo <task description>`
  - Deadline: `deadline <task description> /by <YYYY-MM-DD HH:MM>`
  - Event: `event <task description> /from <YYYY-MM-DD HH:MM> /to <YYYY-MM-DD HH:MM>`
- Delete task: `delete <task index>`
- Mark task: `mark <task index>`
- Unmark task: `unmark <task index>`
- Find task: `find <keyword>`
- Exit: `bye`

Tasks are saved locally on your machine in `./data/tasks.txt`

## Creating your own Duke! 😄 
If you would like to make a Duke as well, you can get the skeleton code [here](https://github.com/nus-cs2103-AY2223S2/ip). This is how the original `main` method looks like:
```java
public class Duke {
    public static void main(String[] args) {
        String logo = " ____        _        \n"
                + "|  _ \\ _   _| | _____ \n"
                + "| | | | | | | |/ / _ \\\n"
                + "| |_| | |_| |   <  __/\n"
                + "|____/ \\__,_|_|\\_\\___|\n";
        System.out.println("Hello from\n" + logo);
    }
}
```

Commit:
Add Gradle support

venuslimm committed:
complete Level-1

venuslimm committed:
complete Level-2

venuslimm committed:
Add Task class

venuslimm committed:
Refactor code into methods

venuslimm committed:
Add Mark as Done and Unmark as Undone features

venuslimm committed:
Create Deadline, Event, Todo classes

venuslimm committed:
Add the 3 features to support todo, event, deadline tasks

venuslimm committed:
Catch error if there is no user input

venuslimm committed:
Add automatic text UI testing

venuslimm committed:
Handle invalid input errors

venuslimm committed:
Add Delete feature

venuslimm committed:
Update javadocs

venuslimm committed:
Handle error when invalid input is specified by user

venuslimm committed:
Refactor code to use enum

venuslimm committed:
Handle invalid input using DukeException

venuslimm committed:
Update text UI test expected file

venuslimm committed:
Reorder message sequence for task deletion

venuslimm committed:
Unindent cases to follow module's coding standards

venuslimm committed:
Modify code comments to follow the module's standards

venuslimm committed:
Refactor code and handle invalid input for Deadline and Event

venuslimm committed:
Add Save and Load tasks from local storage feature

venuslimm committed:
Parse datetime input by user

venuslimm committed:
Merge branch 'branch-Level-8'

venuslimm committed:
Revert "Merge branch 'branch-Level-8'"

This reverts commit d1f4b37bf499039a34e1205e4be9213297d8cc68, reversing
changes made to a5f349effe40e0c5f85f9d524b87f3b8e28db92d.

venuslimm committed:
Add data files to be ignored by git

venuslimm committed:
Merge branch 'branch-Level-7'

venuslimm committed:
Redo reverted merge for branch-Level-8

venuslimm committed:
Refactor code to make it more OOP

venuslimm committed:
Update text-ui-test

venuslimm committed:
Fix path for runtest.bat

venuslimm committed:
Refactor code to categorise them into packages

venuslimm committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

venuslimm committed:
Add JUnit tests for Task and Deadline

venuslimm committed:
Fix mainClassName in build.gradile file

venuslimm committed:
Create manifest.mf file

venuslimm committed:
Add JavaDoc comments

venuslimm committed:
Merge branch 'branch-A-JavaDoc'

venuslimm committed:
Refactor code to comply with coding standard

venuslimm committed:
Merge branch 'branch-A-CodingStandard'

venuslimm committed:
Update all duke.duke occurences in messages to Duke

venuslimm committed:
Add Find feature to allow users to find task using keyword

venuslimm committed:
Merge branch 'branch-Level-9'

venuslimm committed:
Refactor code

venuslimm committed:
Fix text-ui-test

venuslimm committed:
Handle input with the symbol ~

venuslimm committed:
Add warning messages for user if the task is expired or expiring soon

venuslimm committed:
Add junit tests for Parser

venuslimm committed:
Configure to use checkstyle with Gradle

venuslimm committed:
Fix majority of the violations flagged by Checkstyle

venuslimm committed:
Merge branch 'branch-A-CheckStyle'

zhacatomn reviewed (commented):
As far as I can tell, your code obeys the coding standard. LGTM!

seanfirefox commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,23 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        super.isExit = true;
```
Not sure if this follows OOP guidelines, may consider using a setter method.

seanfirefox commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,32 @@
+ * Marks task when user input indicates mark.
+ */
+public class MarkCommand extends Command {
+    private final int TASK_INDEX;
```
Consider using ```private static final int TASK_INDEX``` or ```private final int taskIndex``` as this might have caused warnings in checkstyle.

venuslimm replied:
Resolved. Thanks for the suggestion!

seanfirefox commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private void setEventDateTimes(String description) throws DukeException {
+        Parser parser = new Parser();
```
Can consider setting all methods in parser to be static so there is no need to call the constructor.

venuslimm replied:
Thank you for the suggestion! I have updated all the methods in Parser to be static.

seanfirefox commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+     * Marks task as done.
+     */
+    public void mark() {
+        this.isDone = true;
```
Might want to refrain from unnecessary use of ```this```

venuslimm replied:
Thanks for pointing it out. I have removed the unnecessary use of `this` in some parts of the code.

seanfirefox commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,158 @@
+public class Ui {
+    private final Scanner SCANNER = new Scanner(System.in);
+
+    public String showLine() {
```
Consider changing to a print statement rather than return to follow your naming convention (showWelcome(), showAdd() prints and does not return string)

seanfirefox reviewed (commented):
Overall code quality guidelines are well-followed. Good job!

venuslimm committed:
Change test-based UI to JavaFX UI

venuslimm committed:
Merge branch 'branch-Level-10'

venuslimm committed:
Update JavaFX GUI

venuslimm committed:
Update JavaFX GUI

venuslimm committed:
Allow user to mark and unmark multiple tasks at once using varargs

Mark and Unmark commands: allow user to mark and unmark multiple tasks

Initial implementation of marking and unmarking tasks makes it very
troublesome for users to mark and unmark multiple tasks at once.

Let's use varargs to handle variable number of arguments passed into
mark and unmark methods.

This prevents code duplication.

venuslimm committed:
Merge branch 'branch-A-Varargs'

* branch-A-Varargs:
  Allow user to mark and unmark multiple tasks at once using varargs

venuslimm committed:
Handle non-integer task indexes entered by user

Mark and Unmark commands: skip non-integer task indexes and does not
process them

Code did not handle NumberFormatException, causing program to throw
error.

Let's handle non-integer task indexes by catching NumberFormatException
and processing the next index if there is any.

This ensures that the program does not exit due to exceptions that can
be handled by the programmer.

venuslimm committed:
Add assertions

venuslimm committed:
Merge pull request #1 from venuslimm/branch-A-Assertions

Add assertions

venuslimm committed:
Remove commented and unused codes

venuslimm committed:
Change method used to slice array from Arrays.copyOfRange to Streams

venuslimm committed:
Clean code based on the suggestions from peer review

Code does not include unnecessary keywords like 'this' and static
methods are used.

Code had many redundant keywords, making it hard to read and debug.

Let's
* Remove the redundant 'this' keywords and only using it in the
constructors where there are duplicated variable names.
* Use 'static' methods instead of creating new instances of Parser
every time Parser methods are used.

This cleans up the code and makes it easier to debug.

venuslimm committed:
Modify code to fix checkstyle errors

venuslimm committed:
Refactor code to follow code quality

venuslimm committed:
Merge branch 'master' into branch-A-CodeQuality

venuslimm committed:
Merge pull request #2 from venuslimm/branch-A-CodeQuality

Follow Code Quality

venuslimm committed:
Merge branch 'master' of https://github.com/venuslimm/ip

* 'master' of https://github.com/venuslimm/ip:
  Refactor code to follow code quality
  Modify code to fix checkstyle errors
  Clean code based on the suggestions from peer review

venuslimm committed:
Merge branch 'master' into branch-A-Streams

venuslimm committed:
Merge pull request #3 from venuslimm/branch-A-Streams

Change method used to slice array from Arrays.copyOfRange to Streams

venuslimm committed:
Merge branch 'master' of https://github.com/venuslimm/ip

* 'master' of https://github.com/venuslimm/ip:
  Change method used to slice array from Arrays.copyOfRange to Streams

venuslimm committed:
Modify code to comply to code quality

venuslimm committed:
Fix invalid input length error handling bug for both user and file input

venuslimm committed:
Handle deletion of multiple tasks

venuslimm committed:
Modify output when nothing is marked or unmarked

venuslimm committed:
Add sort feature to sort deadlines or events

venuslimm committed:
Merge pull request #4 from venuslimm/branch-BCD-Extension

Add Sort feature to sort deadlines and events

venuslimm committed:
Refactor code to follow SLAP and pass the checkstyle

venuslimm committed:
Refactor code to follow SLAP and pass checkstyle

venuslimm committed:
Setup CI using GitHub Actions

venuslimm committed:
Add EventTest and edit other JUnit tests to pass CI checks

venuslimm committed:
Fix gradle.yml

venuslimm committed:
Fix header comments for MarkCommand and UnmarkCommand to improve on code
quality

venuslimm committed:
Add credits for using online images

venuslimm committed:
Clean up UI

venuslimm committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Clean up UI

venuslimm committed:
Clean up code and comments

venuslimm committed:
Add tests for FileInputParser and UserInputParser

venuslimm committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  Add tests for FileInputParser and UserInputParser

venuslimm committed:
Change font

venuslimm committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Change font

# Conflicts:
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/ui/Ui.java

venuslimm committed:
Add whitespace when printing on GUI

venuslimm committed:
Add application icon

venuslimm committed:
Edit FXML files so the lengths of the lines are within the valid range

venuslimm committed:
Add representative screenshot of Fake Duke

venuslimm committed:
Add separator after each task when Fake Duke replies to Delete, Mark,
and Unmark commands

venuslimm committed:
Fix grammar when printing on GUI

venuslimm committed:
Add a brief User Guide

venuslimm committed:
Merge branch 'branch-A-UserGuide'

* branch-A-UserGuide:
  Add a brief User Guide

venuslimm committed:
Edit comments and code

venuslimm committed:
Update README.md

venuslimm committed:
Allow future datetimes for Events and Deadlines and the end datetime must be after the start datetime for Events

venuslimm committed:
Handle tasks with empty descriptions

venuslimm committed:
Update README.md

venuslimm committed:
Update README.md to add path of local storage

venuslimm committed:
Remove AM/PM when printing datetime and update error message for find and sort

venuslimm committed:
Update tests as AM/PM are not printed along with datetime on GUI

venuslimm committed:
Update README.md to remove AM/PM

venuslimm committed:
Update new Ui.png

venuslimm committed:
Update Ui.png
